src/index.ts

```diff
@@ -200,6 +210,78 @@ class MayanRouteBase<N extends Network> extends routes.AutomaticRoute<
       : getNativeContractAddress(tokenId.chain);
   }
 
+  async injectReferralInstructionsForSolana(
+    connection: Connection,
+    request: routes.RouteTransferRequest<N>,
+    instructions: TransactionInstruction[],
+    sender: PublicKey,
+    quote: MayanQuote,
+  ) {
+    const { fromChain } = request;
+    const referrerAddress = this.referrerAddress()?.solana;
+    const referralFee = this.getReferralParameters(request).referrerBps;
+
+    if (!referrerAddress || !referralFee || fromChain.network !== 'Mainnet') {
+      return instructions;
+    }
+
+    const referrer = new PublicKey(referrerAddress);
+    const finalInstructions: TransactionInstruction[] = [];
+    const mint = new PublicKey(quote.fromToken.mint);
+    const referrerAta = getAssociatedTokenAddressSync(mint, referrer, true);
+    const senderAta = getAssociatedTokenAddressSync(mint, sender, true);
+    const referrerAtaAccount = await connection.getAccountInfo(referrerAta);
+
+    if (!referrerAtaAccount) {
+      console.error('creating referrer ata');
+      finalInstructions.push(
+        createAssociatedTokenAccountIdempotentInstruction(
+          sender,
+          referrerAta,
+          referrer,
+          mint,
+        ),
+      );
+    }
+
+    finalInstructions.push(
+      createTransferInstruction(senderAta, referrerAta, sender, referralFee),
+    );
+
+    finalInstructions.push(...instructions);
+
+    return instructions;
+  }
+
+  getQuoteAmount(request: routes.RouteTransferRequest<N>, params: Vp) {
+    const referralParams = this.getReferralParameters(request);
+    const referrerBps = referralParams.referrerBps;
+    const decimals = request.source.decimals;
+    const amt = amount.parse(params.amount, decimals);
+
+    // TODO remove solana check when all chains ready
+    const isSolana = request.fromChain.chain === 'Solana';
+
+    if (!referrerBps || !isSolana) {
+      return amt.amount;
+    }
+
+    const dBps = BigInt(10 * referrerBps);
+
+    // https://docs.mayan.finance/integration/referral
```

Comment:
what's the relation to this doc? also I don't think our max is 50, it's limited to `u16` or something. I'm not sure where that's checked in this codebase